File: source/ch/cyberduck/core/serializer/impl/TransferPlistReader.java
package ch.cyberduck.core.serializer.impl;

/*
 * Copyright (c) 2009 David Kocher. All rights reserved.
 * http://cyberduck.ch/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * dkocher@cyberduck.ch
 */

import ch.cyberduck.core.TransferReaderFactory;
import ch.cyberduck.core.serializer.Reader;
import ch.cyberduck.core.transfer.CopyTransfer;
import ch.cyberduck.core.transfer.DownloadTransfer;
import ch.cyberduck.core.transfer.SyncTransfer;
import ch.cyberduck.core.transfer.Transfer;
import ch.cyberduck.core.transfer.UploadTransfer;
import ch.cyberduck.ui.cocoa.foundation.NSDictionary;
import ch.cyberduck.ui.cocoa.foundation.NSObject;

/**
 * @version $Id$
 */
public class TransferPlistReader extends PlistReader<Transfer> {

    public static void register() {
        TransferReaderFactory.addFactory(Factory.NATIVE_PLATFORM, new Factory());
    }

    private static class Factory extends TransferReaderFactory {
        @Override
        public Reader<Transfer> create() {
            return new TransferPlistReader();
        }
    }

    @Override
    public Transfer deserialize(final NSDictionary dict) {
        NSObject kindObj = dict.objectForKey("Kind");
        if(kindObj != null) {
            switch(Transfer.Type.values()[Integer.parseInt(kindObj.toString())]) {
                case download:
                    return new DownloadTransfer(dict);
                case upload:
                    return new UploadTransfer(dict);
                case sync:
                    return new SyncTransfer(dict);
                case copy:
                    return new CopyTransfer(dict);
            }
        }
        throw new IllegalArgumentException("Unknown transfer");
    }
}

File: source/ch/cyberduck/core/serializer/impl/TransferPlistReader.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using System;
using System.Xml;
using ch.cyberduck.core;
using ch.cyberduck.core.transfer;
using org.apache.log4j;

namespace Ch.Cyberduck.Core.Serializer.Impl
{
    public class TransferPlistReader : PlistReader<Transfer>
    {
        private static readonly Logger Log = Logger.getLogger(typeof (TransferPlistReader).FullName);

        public static void Register()
        {
            TransferReaderFactory.addFactory(ch.cyberduck.core.Factory.NATIVE_PLATFORM, new Factory());
        }

        public override Transfer deserialize(XmlNode dictNode)
        {
            XmlNode transferDict = dictNode.SelectSingleNode("dict");
            if (null != transferDict)
            {
                XmlNode kindKeyNode = dictNode.SelectSingleNode("key[.='Kind']");
                if (null != kindKeyNode)
                {
                    int kind = int.Parse(kindKeyNode.NextSibling.InnerText);
                    Transfer.Type type = Transfer.Type.values()[kind];
                    if (type == Transfer.Type.download)
                    {
                        return new DownloadTransfer(dictNode);
                    }
                    if (type == Transfer.Type.upload)
                    {
                        return new UploadTransfer(dictNode);
                    }
                    if (type == Transfer.Type.sync)
                    {
                        return new SyncTransfer(dictNode);
                    }
                    if (type == Transfer.Type.copy)
                    {
                        return new CopyTransfer(dictNode);
                    }
                    Log.error("Unknown transfer:" + kind);
                    throw new ArgumentException("Unknown transfer");
                }
            }
            return null;
        }

        private class Factory : TransferReaderFactory
        {
            protected override object create()
            {
                return new TransferPlistReader();
            }
        }
    }
}